File: web_project/forecaster.js
/**
 * Módulo de Pronóstico - Genera predicciones de series temporales usando ARIMA
 */

class Forecaster {
    constructor() {
        this.forecasts = {};
        this.arimaModels = {};
    }

    /**
     * Genera pronóstico para una categoría usando ARIMA
     */
    forecast(category, weeks = 14) {
        const data = dataProcessor.getWeeklyData(category);
        if (!data) return null;

        const values = data.values;
        const dates = data.dates;

        try {
            // Intentar usar ARIMA
            return this._forecastARIMA(category, values, dates, weeks);
        } catch (e) {
            console.warn(`ARIMA falló para ${category}: ${e.message}. Usando regresión lineal.`);
            // Fallback a regresión lineal
            return this._forecastLinear(category, values, dates, weeks);
        }
    }

    /**
     * Genera pronóstico usando ARIMA
     */
    _forecastARIMA(category, values, dates, weeks) {
        // Encontrar parámetros ARIMA óptimos
        const params = findOptimalARIMAParams(values, 2, 1, 2);

        // Crear y entrenar modelo ARIMA
        const model = new ARIMAModel(values, params);
        model.fit();

        // Generar pronóstico
        const forecastValues = model.forecast(weeks);

        // Calcular intervalos de confianza
        const confidence = model.getConfidenceIntervals(forecastValues);

        // Generar fechas futuras
        const forecastDates = [];
        const lastDate = new Date(dates[dates.length - 1]);

        for (let i = 1; i <= weeks; i++) {
            const futureDate = new Date(lastDate);
            futureDate.setDate(futureDate.getDate() + (i * 7));
            forecastDates.push(futureDate.toISOString().split('T')[0]);
        }

        // Almacenar modelo
        this.arimaModels[category] = { model, params };

        return {
            category: category,
            historicalDates: dates,
            historicalValues: values,
            forecastDates: forecastDates,
            forecastValues: forecastValues,
            confidence: confidence,
            method: 'ARIMA',
            arimaParams: params
        };
    }

    /**
     * Fallback: Genera pronóstico usando regresión lineal
     */
    _forecastLinear(category, values, dates, weeks) {
        const trend = this._calculateTrend(values);
        const seasonal = this._calculateSeasonal(values);

        // Generar pronóstico
        const forecastValues = [];
        const forecastDates = [];

        const lastDate = new Date(dates[dates.length - 1]);

        for (let i = 1; i <= weeks; i++) {
            const futureDate = new Date(lastDate);
            futureDate.setDate(futureDate.getDate() + (i * 7));

            const trendValue = trend.slope * (values.length + i) + trend.intercept;
            const seasonalValue = seasonal[i % seasonal.length];
            const forecastValue = Math.max(0, trendValue * (1 + seasonalValue));

            forecastValues.push(forecastValue);
            forecastDates.push(futureDate.toISOString().split('T')[0]);
        }

        // Calcular intervalos de confianza
        const confidence = this._calculateConfidence(values, forecastValues);

        return {
            category: category,
            historicalDates: dates,
            historicalValues: values,
            forecastDates: forecastDates,
            forecastValues: forecastValues,
            confidence: confidence,
            method: 'Linear Regression'
        };
    }
    
    /**
     * Genera pronósticos para todas las categorías
     */
    forecastAll(weeks = 14) {
        this.forecasts = {};
        
        dataProcessor.categories.forEach(category => {
            this.forecasts[category] = this.forecast(category, weeks);
        });
        
        return this.forecasts;
    }
    
    /**
     * Calcula tendencia usando regresión lineal
     */
    _calculateTrend(values) {
        const n = values.length;
        const x = Array.from({length: n}, (_, i) => i);
        const y = values;
        
        const xMean = x.reduce((a, b) => a + b) / n;
        const yMean = y.reduce((a, b) => a + b) / n;
        
        const numerator = x.reduce((sum, xi, i) => sum + (xi - xMean) * (y[i] - yMean), 0);
        const denominator = x.reduce((sum, xi) => sum + Math.pow(xi - xMean, 2), 0);
        
        const slope = denominator !== 0 ? numerator / denominator : 0;
        const intercept = yMean - slope * xMean;
        
        return { slope, intercept };
    }
    
    /**
     * Calcula componente estacional
     */
    _calculateSeasonal(values) {
        if (values.length < 4) {
            return [0, 0, 0, 0];
        }
        
        const seasonal = [];
        const period = Math.min(4, Math.floor(values.length / 4));
        
        for (let i = 0; i < period; i++) {
            let sum = 0;
            let count = 0;
            
            for (let j = i; j < values.length; j += period) {
                sum += values[j];
                count++;
            }
            
            const avg = sum / count;
            const mean = values.reduce((a, b) => a + b) / values.length;
            seasonal.push((avg - mean) / mean);
        }
        
        return seasonal;
    }
    
    /**
     * Calcula intervalos de confianza
     */
    _calculateConfidence(historical, forecast) {
        const residuals = this._calculateResiduals(historical);
        const std = Math.sqrt(residuals.reduce((sum, r) => sum + r * r, 0) / residuals.length);
        
        const upper = forecast.map(f => f + 1.96 * std);
        const lower = forecast.map(f => Math.max(0, f - 1.96 * std));
        
        return { upper, lower };
    }
    
    /**
     * Calcula residuales
     */
    _calculateResiduals(values) {
        const trend = this._calculateTrend(values);
        return values.map((v, i) => v - (trend.slope * i + trend.intercept));
    }
    
    /**
     * Exporta pronósticos a CSV
     */
    exportToCSV() {
        if (Object.keys(this.forecasts).length === 0) {
            return null;
        }
        
        const firstForecast = Object.values(this.forecasts)[0];
        const dates = firstForecast.forecastDates;
        
        let csv = 'fecha,' + Object.keys(this.forecasts).join(',') + '\n';
        
        dates.forEach((date, idx) => {
            let row = date;
            Object.values(this.forecasts).forEach(forecast => {
                row += ',' + forecast.forecastValues[idx].toFixed(2);
            });
            csv += row + '\n';
        });
        
        return csv;
    }
    
    /**
     * Descarga CSV
     */
    downloadCSV() {
        const csv = this.exportToCSV();
        if (!csv) return;
        
        const blob = new Blob([csv], { type: 'text/csv' });
        const url = window.URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'forecast_' + new Date().toISOString().split('T')[0] + '.csv';
        a.click();
        window.URL.revokeObjectURL(url);
    }
}

// Crear instancia global
const forecaster = new Forecaster();

File: python_project/forecaster.py
"""
Módulo de Pronóstico - Genera predicciones de series temporales
"""

import numpy as np
import pandas as pd
from datetime import datetime, timedelta
from sklearn.linear_model import LinearRegression
from data_processor import DataProcessor

class Forecaster:
    """Genera pronósticos para categorías de consumo"""
    
    def __init__(self, data_path='../data'):
        self.processor = DataProcessor(data_path)
        self.models = {}
    
    def forecast(self, category, weeks=14):
        """Genera pronóstico para una categoría"""
        data = self.processor.get_weekly_data(category)
        
        if 'error' in data:
            return data
        
        dates = pd.to_datetime(data['dates'])
        values = np.array(data['values']).reshape(-1, 1)
        
        # Crear características (índice temporal)
        X = np.arange(len(values)).reshape(-1, 1)
        
        # Entrenar modelo simple (regresión lineal + tendencia)
        model = LinearRegression()
        model.fit(X, values)
        
        # Generar pronóstico
        future_X = np.arange(len(values), len(values) + weeks).reshape(-1, 1)
        forecast_values = model.predict(future_X)
        
        # Agregar componente estacional
        forecast_values = self._add_seasonality(forecast_values, values)
        
        # Generar fechas futuras
        last_date = dates[-1]
        forecast_dates = [
            (last_date + timedelta(weeks=i+1)).strftime('%Y-%m-%d')
            for i in range(weeks)
        ]
        
        return {
            'category': category,
            'forecast_dates': forecast_dates,
            'forecast_values': forecast_values.flatten().tolist(),
            'historical_dates': data['dates'],
            'historical_values': data['values'],
            'confidence_interval': self._calculate_confidence(forecast_values, values)
        }
    
    def forecast_all(self, weeks=14):
        """Genera pronósticos para todas las categorías"""
        categories = self.processor.get_categories()
        forecasts = {}
        
        for category in categories:
            forecasts[category] = self.forecast(category, weeks)
        
        return forecasts
    
    def _add_seasonality(self, forecast, historical):
        """Agrega componente estacional al pronóstico"""
        if len(historical) < 4:
            return forecast
        
        # Calcular promedio móvil para detectar tendencia
        trend = np.convolve(historical.flatten(), np.ones(4)/4, mode='valid')
        
        # Calcular desviación estacional
        seasonal_factor = 1.0 + (np.std(historical) / np.mean(historical)) * 0.1
        
        return forecast * seasonal_factor
    
    def _calculate_confidence(self, forecast, historical):
        """Calcula intervalos de confianza"""
        residuals = np.std(historical)
        
        upper = (forecast + 1.96 * residuals).flatten().tolist()
        lower = (forecast - 1.96 * residuals).flatten().tolist()
        lower = [max(0, x) for x in lower]  # No valores negativos
        
        return {
            'upper': upper,
            'lower': lower
        }
    
    def export_forecast_csv(self, forecasts, filename='forecast.csv'):
        """Exporta pronósticos a CSV"""
        # Preparar datos
        forecast_dates = forecasts[list(forecasts.keys())[0]]['forecast_dates']
        
        data = {'fecha': forecast_dates}
        for category, forecast in forecasts.items():
            data[category] = forecast['forecast_values']
        
        df = pd.DataFrame(data)
        df.to_csv(filename, index=False)
        
        return filename

